File: clients/JavaScript/5.1 mcp_edit_user/MCPEditUserClient.js
const net = require('net');
const readline = require('readline');
const { argv } = require('process');

/**
 * Funktion zum Parsen der Kommandozeilenargumente
 * @param {string[]} args - Array von Kommandozeilenargumenten
 * @returns {Object} - Objekt mit geparsten Argumenten
 */
function parseArguments(args) {
    const parsedArgs = {};
    for (let i = 2; i < args.length; i++) {
        switch (args[i]) {
            case '--server-ip':
                parsedArgs.serverIp = args[++i];
                break;
            case '--server-port':
                parsedArgs.serverPort = parseInt(args[++i], 10);
                break;
            case '--token':
                parsedArgs.token = args[++i];
                break;
            case '--name':
                parsedArgs.name = args[++i];
                break;
            case '--email':
                parsedArgs.email = args[++i];
                break;
            case '--password':
                parsedArgs.password = args[++i];
                break;
            case '--language':
                parsedArgs.language = args[++i];
                break;
            case '--timezone':
                parsedArgs.timezone = args[++i];
                break;
            case '--roles':
                parsedArgs.roles = [];
                while (i + 1 < args.length && !args[i + 1].startsWith('--')) {
                    parsedArgs.roles.push(args[++i]);
                }
                break;
            case '--groups':
                parsedArgs.groups = [];
                while (i + 1 < args.length && !args[i + 1].startsWith('--')) {
                    parsedArgs.groups.push(args[++i]);
                }
                break;
            case '--usePublic':
                parsedArgs.usePublic = true;
                break;
            case '--activateFtp':
                parsedArgs.activateFtp = true;
                break;
            case '--ftpPassword':
                parsedArgs.ftpPassword = args[++i];
                break;
            default:
                console.warn(`⚠️ Unbekanntes Argument: ${args[i]}`);
        }
    }
    return parsedArgs;
}

/**
 * Sendet eine Anfrage an den MCP-Server, um einen Benutzer zu bearbeiten.
 *
 * @param {string} serverIp - IP-Adresse des MCP-Servers
 * @param {number} serverPort - Portnummer des MCP-Servers
 * @param {string} token - Authentifizierungstoken
 * @param {Object} args - Argumente für den zu bearbeitenden Benutzer
 * @returns {Promise<string>} - Antwort vom Server
 */
function sendEditUserRequest(serverIp, serverPort, token, args) {
    return new Promise((resolve, reject) => {
        const client = new net.Socket();
        const payload = {
            command: "edit_user",
            token: token,
            arguments: {
                name: args.name,
                email: args.email,
                password: args.password,
                language: args.language,
                timezone: args.timezone,
                roles: args.roles || [],
                groups: args.groups || [],
                usePublic: args.usePublic || false,
                activateFtp: args.activateFtp || false,
                ftpPassword: args.ftpPassword
            }
        };

        // Entferne Null- oder undefined-Werte
        payload.arguments = Object.fromEntries(Object.entries(payload.arguments).filter(([_, v]) => v != null));

        const payloadString = JSON.stringify(payload);

        // Timeout setzen
        const TIMEOUT_DURATION = 10000; // 10 Sekunden
        const timeout = setTimeout(() => {
            client.destroy();
            reject(new Error('Verbindungs-Timeout: Der Server hat nicht rechtzeitig geantwortet.'));
        }, TIMEOUT_DURATION);

        client.connect(serverPort, serverIp, () => {
            console.log(`🔗 Verbindung zum Server (${serverIp}:${serverPort}) hergestellt.`);
            console.log(`📤 Sende Payload: ${payloadString}`);
            client.write(payloadString);
        });

        let responseData = '';

        client.on('data', (data) => {
            responseData += data.toString();
            try {
                const parsedData = JSON.parse(responseData);
                clearTimeout(timeout);
                resolve(parsedData);
                client.destroy();
            } catch (e) {
                // Weiter empfangen, falls JSON unvollständig ist
            }
        });

        client.on('close', () => {
            console.log('🔒 Verbindung zum Server geschlossen.');
        });

        client.on('error', (err) => {
            clearTimeout(timeout);
            reject(err);
        });
    });
}

// Hauptfunktion
async function main() {
    const args = parseArguments(argv);

    if (!args.serverIp || !args.serverPort || !args.token) {
        console.error('❌ Fehler: --server-ip, --server-port und --token sind erforderlich.');
        console.log('📖 Beispiel: node MCPEditUserClient.js --server-ip 192.168.0.1 --server-port 5000 --token YOUR_AUTH_TOKEN');
        process.exit(1);
    }

    try {
        console.log('🧑‍💻 Sende Edit-User-Anfrage...');
        const response = await sendEditUserRequest(
            args.serverIp,
            args.serverPort,
            args.token,
            args
        );
        console.log('✔️ Antwort vom Server:', JSON.stringify(response, null, 2));
    } catch (err) {
        console.error('❌ Fehler beim Bearbeiten des Benutzers:', err.message);
    }
}

main();

File: clients/Java/5.1 mcp_edit_user/MCPEditUserClient.java
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.*;

public class MCPEditUserClient {
    public static void main(String[] args) {
        // Kurze Validierung der Argumente
        if (args.length < 8) {
            System.out.println("Usage: " +
                    "--server-ip <IP> " +
                    "--server-port <PORT> " +
                    "--token <TOKEN> " +
                    "--user-id <USER_ID> [weitere optionale Parameter]");
            return;
        }

        // Argumente auslesen
        String serverIp   = getArgument(args, "--server-ip");
        String portString = getArgument(args, "--server-port");
        if (serverIp == null || portString == null) {
            System.err.println("Server-IP oder Server-Port fehlt!");
            return;
        }
        int serverPort    = Integer.parseInt(portString);
        String token      = getArgument(args, "--token");
        String userId     = getArgument(args, "--user-id");

        // Optional
        String name       = getArgument(args, "--name");
        String email      = getArgument(args, "--email");
        String password   = getArgument(args, "--password");
        String language   = getArgument(args, "--language");
        String timezone   = getArgument(args, "--timezone");

        List<String> roles  = getListArgument(args, "--roles");
        List<String> groups = getListArgument(args, "--groups");

        boolean usePublic   = Arrays.asList(args).contains("--usePublic");
        boolean activateFtp = Arrays.asList(args).contains("--activateFtp");
        String ftpPassword  = getArgument(args, "--ftpPassword");

        // Argumente in eine Map packen
        Map<String, Object> argumentsMap = new HashMap<>();
        argumentsMap.put("userId",      userId);
        argumentsMap.put("name",        name);
        argumentsMap.put("email",       email);
        argumentsMap.put("password",    password);
        argumentsMap.put("language",    language);
        argumentsMap.put("timezone",    timezone);
        argumentsMap.put("roles",       roles);
        argumentsMap.put("groups",      groups);
        argumentsMap.put("usePublic",   usePublic);
        argumentsMap.put("activateFtp", activateFtp);
        argumentsMap.put("ftpPassword", ftpPassword);

        // Null oder leere Strings entfernen
        argumentsMap = removeNullOrEmpty(argumentsMap);

        // JSON-Payload zusammenstellen
        JSONObject payload = new JSONObject();
        payload.put("command", "edit_user");
        payload.put("token", token);
        payload.put("arguments", argumentsMap);

        System.out.println("📤 Sending edit user request...");
        String response = sendRequest(serverIp, serverPort, payload.toString());
        System.out.println("✔️ Response from server:");
        System.out.println(response);
    }

    /**
     * Liest den Wert eines bestimmten Parameters aus dem Argument-Array.
     * Bsp.: "--server-ip" -> "127.0.0.1"
     *
     * @param args Array mit allen Argumenten
     * @param key  Name des Arguments, nach dem gesucht wird (z.B. "--server-ip")
     * @return     Wert des Arguments oder null, falls nicht gefunden
     */
    private static String getArgument(String[] args, String key) {
        for (int i = 0; i < args.length; i++) {
            if (args[i].equals(key) && i < args.length - 1) {
                return args[i + 1];
            }
        }
        return null;
    }

    /**
     * Liest mehrere Werte für einen Argument-Schlüssel (z.B. "--roles") aus,
     * bis ein neues Argument mit "--" beginnt.
     *
     * @param args Array mit allen Argumenten
     * @param key  Name des Arguments, nach dem gesucht wird (z.B. "--roles")
     * @return     Liste der gefundenen Werte, sonst eine leere Liste
     */
    private static List<String> getListArgument(String[] args, String key) {
        List<String> values = new ArrayList<>();
        for (int i = 0; i < args.length; i++) {
            if (args[i].equals(key)) {
                // Sammle solange, bis ein neues "--" beginnt oder args zu Ende
                for (int j = i + 1; j < args.length && !args[j].startsWith("--"); j++) {
                    values.add(args[j]);
                }
                break;
            }
        }
        return values;
    }

    /**
     * Entfernt alle Einträge, deren Wert null oder ein leerer String ist.
     */
    private static Map<String, Object> removeNullOrEmpty(Map<String, Object> original) {
        Map<String, Object> cleaned = new HashMap<>();
        for (Map.Entry<String, Object> entry : original.entrySet()) {
            Object value = entry.getValue();
            if (value != null) {
                // Wenn String, checke ob leer oder nur Whitespace
                if (value instanceof String) {
                    String str = (String) value;
                    if (!str.trim().isEmpty()) {
                        cleaned.put(entry.getKey(), value);
                    }
                } else {
                    // Keine weitere Prüfung für Listen, Booleans, etc.
                    cleaned.put(entry.getKey(), value);
                }
            }
        }
        return cleaned;
    }

    /**
     * Sendet den gegebenen JSON-String per TCP an den Server und empfängt die Antwort.
     * 
     * @param serverIp   IP des Servers
     * @param serverPort Port des Servers
     * @param payload    JSON-String
     * @return           Antwort als String oder Fehlermeldung
     */
    private static String sendRequest(String serverIp, int serverPort, String payload) {
        try (Socket socket = new Socket(serverIp, serverPort)) {
            // Daten zum Server senden
            OutputStream output = socket.getOutputStream();
            byte[] data = payload.getBytes(StandardCharsets.UTF_8);
            output.write(data);
            output.flush();

            // Antwort lesen
            BufferedReader reader = new BufferedReader(
                new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8)
            );
            StringBuilder sb = new StringBuilder();
            char[] buffer = new char[4096];
            int charsRead;
            do {
                charsRead = reader.read(buffer, 0, buffer.length);
                if (charsRead > 0) {
                    sb.append(buffer, 0, charsRead);
                }
            } while (charsRead == buffer.length);

            return sb.toString();

        } catch (Exception e) {
            return "Error: " + e.getMessage();
        }
    }
}
